rllib/core/optim/marl_optimizer.py

```diff
@@ -0,0 +1,152 @@
+            optim_kwargs: The named mapping of optimizer configs to use for
+
+        Returns:
+            A MultiAgentRLOptimizer.
+        """
+        assert len(rl_optimizer_classes) == len(optim_kwargs)
+        assert set(rl_optimizer_classes.keys()) == set(optim_kwargs.keys())
+        optimizers = {}
+        for module_id in module.keys():
+            submodule = module[module_id]
+            config = optim_kwargs[module_id]
+            optimizers[module_id] = rl_optimizer_classes[module_id].from_module(
+                submodule, config
+            )
+        return cls(optimizers)
+
+    @override(RLOptimizer)
+    def compute_loss(
+        self,
+        fwd_out: Mapping[ModuleID, Mapping[str, Any]],
+        batch: Mapping[ModuleID, Mapping[str, Any]],
+    ) -> Union[TensorType, Mapping[str, Any]]:
+        total_loss = None
+        ret = {}
+        for module_id in batch.keys():
+            optimizer = self._optimizers[module_id]
+            loss = optimizer.compute_loss(fwd_out[module_id], batch[module_id])
+            if not isinstance(loss, collections_mapping):
+                loss = {"total_loss": loss}
+            if total_loss is None:
+                total_loss = loss["total_loss"]
+            else:
+                total_loss += loss["total_loss"]
+            ret[module_id] = loss
+        ret["total_loss"] = total_loss
+        return ret
+
+    @override(RLOptimizer)
+    def get_state(self) -> Mapping[ModuleID, Mapping[str, Any]]:
+        """Get the optimizer state.
+
+        Returns:
+            The optimizer state.
+        """
+        return {
+            module_id: optimizer.get_state()
+            for module_id, optimizer in self._optimizers.items()
+        }
+
+    @override(RLOptimizer)
+    def set_state(self, state: Mapping[ModuleID, Mapping[str, Any]]) -> None:
+        """Set the optimizer state.
+
+        Args:
+            state: The optimizer state to set.
+        """
+        for module_id, sub_optimizer_state in state.items():
+            optimizer = self._optimizers[module_id]
+            optimizer.set_state(sub_optimizer_state)
+
+    @override(RLOptimizer)
+    def as_multi_agent(self) -> "MultiAgentRLOptimizer":
+        return self
+
+    def add_optimizer(self, module_id: ModuleID, optimizer: RLOptimizer) -> None:
+        """Add a new optimizer to the multi-agent optimizer.
+
+        Args:
+            module_id: The module id of the optimizer.
+            optimizer: The optimizer to add.
+        """
+        self._optimizers[module_id] = optimizer
+
+    def remove_optimizer(self, module_id: ModuleID) -> None:
+        """Remove an optimizer from the multi-agent optimizer.
+
+        Args:
+            module_id: The module id of the optimizer to remove.
+        """
+        del self._optimizers[module_id]
+
+    def _configure_optimizers(self) -> None:
+        # Do not override this.
+        return self._optimizers
```

Comment:
this is a hack that I don't love that's required since we're using @kouroshHakha 's init_subclass decorator trick.

I tried overriding init_subclass in this class to remove the logic from RLOptimizer, but for whatever reason that didn't work. It lead to the super class's init_subclass method getting called still. 

I'll come back and try to fix this, but in the meantime it's not important since this function shouldn't be modified by users anyways.

Reply:
discussed the issue offline. will resolve in the marl optimizer pr